File: kotlin-node/karakum/src/jsMain/resources/nameResolvers/resolvePerformanceObserverOptionsName.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

export default (node, context) => {
    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"
    if (!sourceFileName.endsWith("perf_hooks.d.ts")) return null

    const typeScriptService = context.lookupService(karakum.typeScriptServiceKey.get())
    const getParent = typeScriptService?.getParent.bind(typeScriptService) ?? (node => node.parent)

    const parameterType = getParent(node)
    if (!parameterType) return null
    if (!ts.isUnionTypeNode(parameterType)) return null

    let infix = "Type"
    if (
        ts.isTypeLiteralNode(node)
        && node.members.some(member => (
            ts.isPropertySignature(member)
            && ts.isIdentifier(member.name)
            && member.name.text === "entryTypes"
        ))
    ) {
        infix = "Types"
    }

    const parameter = getParent(parameterType)
    if (!parameter) return null
    if (!ts.isParameter(parameter)) return null
    if (!ts.isIdentifier(parameter.name)) return null
    if (parameter.name.text !== "options") return null

    const method = getParent(parameter)
    if (!method) return null
    if (!ts.isMethodDeclaration(method)) return null
    if (!ts.isIdentifier(method.name)) return null
    if (method.name.text !== "observe") return null

    const classNode = getParent(method)
    if (!classNode) return null
    if (!ts.isClassDeclaration(classNode)) return null
    if (classNode.name === undefined) return null
    if (classNode.name.text !== "PerformanceObserver") return null

    return `PerformanceObserverObserve${infix}Options`
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/nameResolvers/resolvePerformanceObserverOptionsName.kt
package node.karakum.nameResolvers

import io.github.sgrishchenko.karakum.extension.Context
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import arrow.core.raise.nullable
import typescript.Node
import typescript.asArray
import typescript.isClassDeclaration
import typescript.isIdentifier
import typescript.isMethodDeclaration
import typescript.isParameter
import typescript.isPropertySignature
import typescript.isTypeLiteralNode
import typescript.isUnionTypeNode

fun resolvePerformanceObserverOptionsName(node: Node, context: Context) = nullable {
    val typeScriptService = ensureNotNull(context.lookupService<TypeScriptService>(typeScriptServiceKey))

    val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
    ensure(sourceFileName.endsWith("perf_hooks.d.ts"))

    val parameterType = ensureNotNull(typeScriptService.getParent(node))
    ensure(isUnionTypeNode(parameterType))

    val parameter = ensureNotNull(typeScriptService.getParent(parameterType))
    ensure(isParameter(parameter))

    val parameterNameNode = parameter.name
    ensure(isIdentifier(parameterNameNode))
    val parameterName = parameterNameNode.text
    ensure(parameterName == "options")

    val method = ensureNotNull(typeScriptService.getParent(parameter))
    ensure(isMethodDeclaration(method))

    val methodNameNode = method.name
    ensure(isIdentifier(methodNameNode))
    val methodName = methodNameNode.text
    ensure(methodName == "observe")

    val classNode = ensureNotNull(typeScriptService.getParent(method))
    ensure(isClassDeclaration(classNode))
    val parentName = ensureNotNull(classNode.name).text
    ensure(parentName == "PerformanceObserver")

    nullable {
        ensure(isTypeLiteralNode(node))
        ensure(node.members.asArray().any { member ->
            nullable {
                ensure(isPropertySignature(member))

                val propertyName = member.name
                ensure(isIdentifier(propertyName))
                ensure(propertyName.text == "entryTypes")
            } != null
        })

        "PerformanceObserverObserveTypesOptions"
    } ?: run {
        "PerformanceObserverObserveTypeOptions"
    }
}
